### ISSUE TITLE
PGM direct access optimization : performance comparison

### ISSUE THREAD

**Issue by SuperUserNameMan** [dbuezas/lgt8fx#59]
Hello,

Regarding the discussion and benchmarks in #33 about accessing PROGMEM directly, I've done some more benchmarks.

The sketch used as benchmark outputs an analog wave through the pin DAC0.
The wave is stored into an array which is read from SRAM or from PROGMEM using the standard ``pgm_read_byte()`` macro, or different direct access methods.
Then, using an oscilloscope, I measured the frequency of the wave.

Here are the results at 32 MHz :

\# | Storage | Repeat | Access | Output Frequency | Flash usage | RAM usage
--|-----------|-----------|----------------------|---------|-------------|-----
1 | SRAM | inline | table[ adr ] | 503 Hz | 1628 | 283
2 | SRAM | loop | table[ adr ] | 490 Hz | 1586 | 283
3 | PROGMEM | inline | pgm_read_byte( adr ) | 455 Hz | 1662 | 11
4 | PROGMEM | loop | pgm_read_byte( adr ) | 455 Hz | 1588 | 11
5 | PROGMEM | inline |  (\*((uint8_t*)( 0x4000 + adr ))) | 472 Hz | 1636 | 11
6 | PROGMEM | loop | (\*((uint8_t*)( 0x4000 + adr ))) | 472 Hz | 1586 | 11
7 | PROGMEM | inline |  (\*((uint16_t*)( 0x4000 + adr ))) | 455 Hz | 1638 | 11
8 | PROGMEM | loop | (\*((uint16_t*)( 0x4000 + adr ))) | 423 Hz | 1588 | 11
9 | PROGMEM | inline |  table[ 0x4000 + adr ] | 502 Hz | 1628 | 11
10 | PROGMEM | loop | table[ 0x4000 + adr ] | 472 Hz | 1586 | 11

----

So, according to these tests, the fastest way to direct-access PROGMEM is ``table[ 0x4000 + adr ]``, which is almost as fast as SRAM access, which both seems to consume the least flash-space.

----

```` C++

//#define TEST_1_SRAM_10_INLINE           //!\\ <--- don't forget to comment PROGMEM
//#define TEST_2_SRAM_10_LOOP             //!\\ <--- don't forget to comment PROGMEM
//#define TEST_3_PGM_READ_BYTE_10_INLINE
//#define TEST_4_PGM_READ_BYTE_10_LOOP
//#define TEST_5_DIRECT_8bits_10_INLINE
//#define TEST_6_DIRECT_8bits_10_LOOP
//#define TEST_7_DIRECT_16bits_10_INLINE
//#define TEST_8_DIRECT_16bits_10_LOOP
#define TEST_9_DIRECT_OFFSET_10_INLINE
//#define TEST_10_DIRECT_OFFSET_10_LOOP

PROGMEM		//!\\ <--- don't forget to comment when testing SRAM mode
const byte table[] = {
	0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,
	0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,
	0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,
	0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,
	0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,
	0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,
	0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,
	0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,
	
	0xFF,0xFF,0xFF,0xFF,0xFF,0xFF,0xFF,0xFF,0xFF,0xFF,0xFF,0xFF,0xFF,0xFF,0xFF,0xFF,0xFF,
	0xFF,0xFF,0xFF,0xFF,0xFF,0xFF,0xFF,0xFF,0xFF,0xFF,0xFF,0xFF,0xFF,0xFF,0xFF,0xFF,0xFF,
	0xFF,0xFF,0xFF,0xFF,0xFF,0xFF,0xFF,0xFF,0xFF,0xFF,0xFF,0xFF,0xFF,0xFF,0xFF,0xFF,0xFF,
	0xFF,0xFF,0xFF,0xFF,0xFF,0xFF,0xFF,0xFF,0xFF,0xFF,0xFF,0xFF,0xFF,0xFF,0xFF,0xFF,0xFF,
	0xFF,0xFF,0xFF,0xFF,0xFF,0xFF,0xFF,0xFF,0xFF,0xFF,0xFF,0xFF,0xFF,0xFF,0xFF,0xFF,0xFF,
	0xFF,0xFF,0xFF,0xFF,0xFF,0xFF,0xFF,0xFF,0xFF,0xFF,0xFF,0xFF,0xFF,0xFF,0xFF,0xFF,0xFF,
	0xFF,0xFF,0xFF,0xFF,0xFF,0xFF,0xFF,0xFF,0xFF,0xFF,0xFF,0xFF,0xFF,0xFF,0xFF,0xFF,0xFF,
	0xFF,0xFF,0xFF,0xFF,0xFF,0xFF,0xFF,0xFF,0xFF,0xFF,0xFF,0xFF,0xFF,0xFF,0xFF,0xFF,0xFF,
};

#define SAMPLES 254

void setup() 
{
	delay(1000);
	
	analogReference(DEFAULT);
	
	pinMode(DAC0, ANALOG);
}


void loop() 
{
	for (byte i = 0; i <= SAMPLES; i++) {
		
		#if 0
		#elif defined( TEST_1_SRAM_10_INLINE )
			analogWrite( DAC0, table[ i ] );
			analogWrite( DAC0, table[ i ] );
			analogWrite( DAC0, table[ i ] );
			analogWrite( DAC0, table[ i ] );
			analogWrite( DAC0, table[ i ] );
			analogWrite( DAC0, table[ i ] );
			analogWrite( DAC0, table[ i ] );
			analogWrite( DAC0, table[ i ] );
			analogWrite( DAC0, table[ i ] );
			analogWrite( DAC0, table[ i ] );
			
		#elif defined( TEST_2_SRAM_10_LOOP )
			for( byte t=0; t<10; t++ ) analogWrite( DAC0, table[ i ] );
			
		#elif defined( TEST_3_PGM_READ_BYTE_10_INLINE )
			analogWrite( DAC0, pgm_read_byte( table + i ) );
			analogWrite( DAC0, pgm_read_byte( table + i ) );
			analogWrite( DAC0, pgm_read_byte( table + i ) );
			analogWrite( DAC0, pgm_read_byte( table + i ) );
			analogWrite( DAC0, pgm_read_byte( table + i ) );
			analogWrite( DAC0, pgm_read_byte( table + i ) );
			analogWrite( DAC0, pgm_read_byte( table + i ) );
			analogWrite( DAC0, pgm_read_byte( table + i ) );
			analogWrite( DAC0, pgm_read_byte( table + i ) );
			analogWrite( DAC0, pgm_read_byte( table + i ) );
			
		#elif defined( TEST_4_PGM_READ_BYTE_10_LOOP )
			for( byte t=0; t<10; t++ ) analogWrite( DAC0, pgm_read_byte( table + i ) );
			
		#elif defined( TEST_5_DIRECT_8bits_10_INLINE )
			analogWrite( DAC0, (*((uint8_t*)( 0x4000 + table + i ))) );
			analogWrite( DAC0, (*((uint8_t*)( 0x4000 + table + i ))) );
			analogWrite( DAC0, (*((uint8_t*)( 0x4000 + table + i ))) );
			analogWrite( DAC0, (*((uint8_t*)( 0x4000 + table + i ))) );
			analogWrite( DAC0, (*((uint8_t*)( 0x4000 + table + i ))) );
			analogWrite( DAC0, (*((uint8_t*)( 0x4000 + table + i ))) );
			analogWrite( DAC0, (*((uint8_t*)( 0x4000 + table + i ))) );
			analogWrite( DAC0, (*((uint8_t*)( 0x4000 + table + i ))) );
			analogWrite( DAC0, (*((uint8_t*)( 0x4000 + table + i ))) );
			analogWrite( DAC0, (*((uint8_t*)( 0x4000 + table + i ))) );
			
		#elif defined( TEST_6_DIRECT_8bits_10_LOOP )
			for( byte t=0; t<10; t++ ) analogWrite( DAC0, (*((uint8_t*)( 0x4000 + table + i ))) );
			
		#elif defined( TEST_7_DIRECT_16bits_10_INLINE )
			analogWrite( DAC0, (*((uint16_t*)( 0x4000 + table + i ))) );
			analogWrite( DAC0, (*((uint16_t*)( 0x4000 + table + i ))) );
			analogWrite( DAC0, (*((uint16_t*)( 0x4000 + table + i ))) );
			analogWrite( DAC0, (*((uint16_t*)( 0x4000 + table + i ))) );
			analogWrite( DAC0, (*((uint16_t*)( 0x4000 + table + i ))) );
			analogWrite( DAC0, (*((uint16_t*)( 0x4000 + table + i ))) );
			analogWrite( DAC0, (*((uint16_t*)( 0x4000 + table + i ))) );
			analogWrite( DAC0, (*((uint16_t*)( 0x4000 + table + i ))) );
			analogWrite( DAC0, (*((uint16_t*)( 0x4000 + table + i ))) );
			analogWrite( DAC0, (*((uint16_t*)( 0x4000 + table + i ))) );
			
		#elif defined( TEST_8_DIRECT_16bits_10_LOOP )
			for( byte t=0; t<10; t++ ) analogWrite( DAC0, (*((uint16_t*)( 0x4000 + table + i ))) );
			
		#elif defined( TEST_9_DIRECT_OFFSET_10_INLINE )
			analogWrite( DAC0, table[ 0x4000 + i ] );
			analogWrite( DAC0, table[ 0x4000 + i ] );
			analogWrite( DAC0, table[ 0x4000 + i ] );
			analogWrite( DAC0, table[ 0x4000 + i ] );
			analogWrite( DAC0, table[ 0x4000 + i ] );
			analogWrite( DAC0, table[ 0x4000 + i ] );
			analogWrite( DAC0, table[ 0x4000 + i ] );
			analogWrite( DAC0, table[ 0x4000 + i ] );
			analogWrite( DAC0, table[ 0x4000 + i ] );
			analogWrite( DAC0, table[ 0x4000 + i ] );
			
		#elif defined( TEST_10_DIRECT_OFFSET_10_LOOP )
			for( byte t=0; t<10; t++ ) analogWrite( DAC0, table[ 0x4000 + i ] );
			
		#endif
	}
}
````

**Comment by SuperUserNameMan** [comment]
ok, i closed this issue temporarily because my LQFP32 board was resetting itself for unknown reasons whith ``TEST_3_PGM_READ_BYTE_10_INLINE``. But it stopped doing that ... so i updated the benchmark above, and reopen the issue.

**Comment by LaZsolt** [comment]
analogWrite is very slow. Try my code to get higher freq. ( Your table contains 272 byte. 17 x 16 byte )

Anyway your test is completely useless. In fact every comparison is useless, because compiler does not know what are we wanted to address at above 0x4000. That's why the compiler generate a less optimized code.

This code below I designed 3 cases. RAM addressing, progmem addressing and remapped progmem addressing.
- In first case compiler does everything perfect when I watch the generated assembly.
- In second case the compiled code is not optimal, because overwrite the address register with readed data. So extra code generater to restore.
- In 3-rd case inside the for cycle the reading address is calculated every times. (Adding 0x4000 to the pointer at every reads) But in C language is written: *((char*)p++)
```
#define TESTTYPE 1

#if   TESTTYPE == 1
  // RAM read
  #define PTR_INIT( _var ) (uint16_t) (&_var)
  #define PGM_READ( _adr ) *((char*)_adr++)
#elif TESTTYPE == 2
  // PROGMEM read
  #define PTR_INIT( _var ) (uint16_t) &_var
  #define PGM_READ( _adr ) pgm_read_byte( _adr++ ) 
#elif TESTTYPE == 3
  // Remapped PROGMEM read
  #define PTR_INIT( _var ) 0x4000 + (uint16_t) (&_var)
  #define PGM_READ( _adr ) *((char*)_adr++)
#endif

PROGMEM    //!\\ <--- don't forget to comment when testing SRAM mode
const byte table[] = {
  0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,
  0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,
  0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,
  0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,
  0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,
  0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,
  0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,
  0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,
  
  0xFF,0xFF,0xFF,0xFF,0xFF,0xFF,0xFF,0xFF,0xFF,0xFF,0xFF,0xFF,0xFF,0xFF,0xFF,0xFF,
  0xFF,0xFF,0xFF,0xFF,0xFF,0xFF,0xFF,0xFF,0xFF,0xFF,0xFF,0xFF,0xFF,0xFF,0xFF,0xFF,
  0xFF,0xFF,0xFF,0xFF,0xFF,0xFF,0xFF,0xFF,0xFF,0xFF,0xFF,0xFF,0xFF,0xFF,0xFF,0xFF,
  0xFF,0xFF,0xFF,0xFF,0xFF,0xFF,0xFF,0xFF,0xFF,0xFF,0xFF,0xFF,0xFF,0xFF,0xFF,0xFF,
  0xFF,0xFF,0xFF,0xFF,0xFF,0xFF,0xFF,0xFF,0xFF,0xFF,0xFF,0xFF,0xFF,0xFF,0xFF,0xFF,
  0xFF,0xFF,0xFF,0xFF,0xFF,0xFF,0xFF,0xFF,0xFF,0xFF,0xFF,0xFF,0xFF,0xFF,0xFF,0xFF,
  0xFF,0xFF,0xFF,0xFF,0xFF,0xFF,0xFF,0xFF,0xFF,0xFF,0xFF,0xFF,0xFF,0xFF,0xFF,0xFF,
  0xFF,0xFF,0xFF,0xFF,0xFF,0xFF,0xFF,0xFF,0xFF,0xFF,0xFF,0xFF,0xFF,0xFF,0xFF,0xFF,
};

#define SAMPLES 255

void setup(){
  analogReference(DEFAULT);
  pinMode(DAC0, ANALOG);
}


void loop(){
  uint16_t p;

while (1) {                                  // Without this while, compiler generate call 0 at the end of for cycle.
      p = PTR_INIT( table );

    for (uint16_t i = 0; i <= SAMPLES; i++) {
      DALR = PGM_READ( p );                  // Write direct into DAC register
    }
  }
}
```

**Comment by dbuezas** [comment]
I'd recommend to disable timer0 to avoid the millis interrupt getting in the way.
```cpp
  bitClear(TIMSK0, TOIE0); 
```

### TESTTYPE = 1 --> 64us = 15,625 Hz
![image](https://user-images.githubusercontent.com/777196/96510238-80608600-125d-11eb-9266-4bd2608e978d.png)

### TESTTYPE = 2 --> 96us = 10,417 Hz
![image](https://user-images.githubusercontent.com/777196/96510267-8bb3b180-125d-11eb-82c1-173ddeb7b07e.png)

### TESTTYPE = 3 --> 104us = 9.615 Hz
![image](https://user-images.githubusercontent.com/777196/96510311-9c642780-125d-11eb-8bbf-b9ea37874084.png)

---

by the way, if you allow me I'd rephrase "**Anyway your test is ~completely useless~**" to "**Anyway your test is skewed by compiler optimisations**"

I personally appreciate the brains and curiosity behind all these experiments! ❤️

**Comment by SuperUserNameMan** [comment]
@LaZsolt :

> ... because compiler does not know what are we wanted to address at above 0x4000. That's why the compiler generate a less optimized code.

Yeah, I don't know yet how to look at the assembly code, so I have no idea what the compiler does behind the curtain. 

In that condition, the only thing I can compare are performances in an imperfect benchmark that I tried to make closer to a real use case of mine.
Direct write to ``DALR`` is a good idea indeed, but using the ``while(1)`` you enclosed into the ``loop()`` or disabling timer0 as proposed by @dbuezas could be incompatible with a real use-case scenario.

Maybe I'm wrong, but even I we manage to write an over-optimized benchmark, we have no clue that the compiler will behave the same way in a random real use case. So, in that perspective, we could say that any benchmark is useless. And that the only useful benchmark is the real use-case that we want to optimize.

BTW, I've noticed that ``for (uint16_t i = 0; i <= SAMPLES; i++)``  was slower than ``for (uint8_t i = 0; i <= SAMPLES; i++) {``, so if our point is to get the maximum output speed, we might want to avoid ``uint16_t`` or ``auto`` for loops and stick to ``uint8_t``, unless of course if the compiled assembly code shows otherwise in a real use-case.

Even if I was aware that it is still not a perfect benchmark, this test shown me that it was possible to access ``PROGMEM`` as ``table[0x4000 + adr]`` and that it could be faster than the other methods we previously tested in #33.

Again, I don't know if the ``table[0x4000 + adr]`` is calculated each time, but in this benchmark I noticed that using ``const uint8_t * pgm_table = 0x4000 + table;`` and ``pgm_table[ adr ]`` is slower. (``table[0x4000 + adr]`` is the fastest).

----

@dbuezas :

> I'd recommend to disable timer0 to avoid the millis interrupt getting in the way.

Oh ! So that was it ! I was wondering why the signal was slightly unstable.

And that's because of this instability that I reduced the frequency by a factor of 10 so I could read an exact frequency (range 0-999) on my DSO213.

----

Maybe I'll come back again later with a better "almost real-use-case scenario" benchmark and share my results with the community.

**Comment by LaZsolt** [comment]
@SuperUserNameMan 

> in this benchmark I noticed that using const uint8_t * pgm_table = 0x4000 + table; and pgm_table[ adr ] is slower. (table[0x4000 + adr] is the fastest).

Strange behaviour. I'll investigate it later.

**Comment by SuperUserNameMan** [comment]
> @SuperUserNameMan
> 
> > in this benchmark I noticed that using const uint8_t * pgm_table = 0x4000 + table; and pgm_table[ adr ] is slower. (table[0x4000 + adr] is the fastest).
> 
> Strange behaviour. I'll investigate it later.

Yes that would be appropriate to investigate further, because this afternoon, I was trying to use ``table[0x4000+adr]`` in an other project, and it was not working at all.

**Comment by SuperUserNameMan** [comment]
Ok, after further tests, I've found that my beloved brand new ``table[0x4000+adr]`` was not reliable, and that it had some strange unexpected behaviors.

Here is a test code that makes it obvious :
```` C++

const uint8_t data[] PROGMEM = { 123, 1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20 };

int data_len = sizeof( data );

const uint8_t * pgm_data = (uint8_t*)((uint16_t)&data + 0x4000);


void setup()
{
	//delay(3000);
	
	Serial.begin(9600);
	
	Serial.println("---"); // ---- display memory adresses 
	
	Serial.print( " data         = 0x" ); Serial.println( (uint16_t) data, HEX );
	Serial.print( "&data         = 0x" ); Serial.println( (uint16_t) &data, HEX );
	Serial.print( "&data[0]      = 0x" ); Serial.println( (uint16_t) &data[0], HEX );
	Serial.print( "&data[0x4000] = 0x" ); Serial.println( (uint16_t) &data[0x4000], HEX );
	Serial.print( " pgm_data     = 0x" ); Serial.println( (uint16_t) pgm_data, HEX );
	Serial.print( "&pgm_data     = 0x" ); Serial.println( (uint16_t) &pgm_data, HEX );
	
	
	Serial.println("---"); // ---- display reference
	
	Serial.print("pgm_read_byte( data ) = ");
	Serial.print( pgm_read_byte( data ) );
	Serial.println();
	
	Serial.print("pgm_read_byte( data + a ) = { ");
	for( int a=0; a < data_len ; a++ )
	{
		Serial.print( pgm_read_byte( data + a ) );
		Serial.print(", ");
	}
	Serial.println(" }");
		
	
	Serial.println("---"); // --- display data[ a ]
	
	Serial.print("data[ 0 ] = ");
	Serial.print( data[ 0 ] );
	Serial.println();
	
	Serial.print("data[ a ] = { ");
	for( int a=0; a < data_len ; a ++ )
	{
		Serial.print( data[ a ] );
		Serial.print(", ");
	}
	Serial.println(" }");
	
	
	Serial.println("---"); // --- display table[ 0x4000 +a ]
	
	Serial.print("data[ 0x4000 ] = ");
	Serial.print( data[ 0x4000 ] );
	Serial.println();
		
	Serial.print("data[ 0x4000 + a ] = { ");
	for( int a=0; a < data_len ; a ++ )
	{
		Serial.print( data[ 0x4000 + a ] );
		Serial.print(", ");
	}
	Serial.println(" }");
			
			
	Serial.println("---"); // --- display pgm_data[ a ]
	
	Serial.print("pgm_data[ 0 ] = ");
	Serial.print( pgm_data[ 0 ] );
	Serial.println();
	
	Serial.print("pgm_data[ a ] = { ");
	for( int a=0; a < data_len ; a ++ )
	{
		Serial.print( pgm_data[ a ] );
		Serial.print(", ");
	}
	Serial.println(" }");
	
		
	Serial.flush();
}

void loop()
{
}

````
----
Unexpected results that i get are :
```
---
 data         = 0x68
&data         = 0x68
&data[0]      = 0x68
&data[0x4000] = 0x4068
 pgm_data     = 0x4068
&pgm_data     = 0x100
---
pgm_read_byte( data ) = 123
pgm_read_byte( data + a ) = { 123, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20,  }
---
data[ 0 ] = 123
data[ a ] = { 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 135, 0, 0,  }
---
data[ 0x4000 ] = 0
data[ 0x4000 + a ] = { 123, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20,  }
---
pgm_data[ 0 ] = 123
pgm_data[ a ] = { 123, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20,  }

```
----
I have no clue why, and it's starting to make me crazy ... is it my brain ? is it GCC ? it it covid ? i don't know.

My config : 
- Linux 5.4.0-52-generic x86_x64
- Arduino 1.8.13 (downloaded from the official web site)
I have not clue which version of GCC it uses.

**Comment by SuperUserNameMan** [comment]
Ok, I think i understood why ``table[0x4000+adr]`` was not reliable.

``` C++
const uint8_t table[] PROGMEM = { 123, 1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20 };
```

``table[ 7 ]`` will work because the compiler will replace it with the constant value from the array.

``table[ 0x4000 + 7 ]`` will NOT work because the compiler will replace with constant 0 because out of bound.

``table[ adr ]`` will not work because the offset is missing.

``table[ 0x4000 + adr ]`` will work.

On the other hand :

``` C++
const uint8_t table[] PROGMEM = { 123, 1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20 };

const uint8_t * pgm_data = (uint8_t*)((uint16_t)&data + 0x4000);
```

``pgm_table[ 7 ]`` will NOT work because the compiler will replace it with the constant 0 because it thinks it is out of bound.

``pgm_table[ adr ]`` will work.

----

Here is a sketch that show this beahavior :

```` C++

const uint8_t data[] PROGMEM = { 123, 1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20 };

int data_len = sizeof( data );

const uint8_t * pgm_data = (uint8_t*)((uint16_t)&table + 0x4000);


void setup()
{
	//delay(3000);
	
	Serial.begin(9600);
	
	Serial.println("---"); // ---- display memory adresses 
	
	Serial.print( " data         = 0x" ); Serial.println( (uint16_t) data, HEX );
	Serial.print( "&data         = 0x" ); Serial.println( (uint16_t) &data, HEX );
	Serial.print( "&data[0]      = 0x" ); Serial.println( (uint16_t) &data[0], HEX );
	Serial.print( "&data[0x4000] = 0x" ); Serial.println( (uint16_t) &data[0x4000], HEX );
	Serial.print( " pgm_data     = 0x" ); Serial.println( (uint16_t) pgm_data, HEX );
	Serial.print( "&pgm_data     = 0x" ); Serial.println( (uint16_t) &pgm_data, HEX );
	
	
	Serial.println("---"); // ---- display reference
	
	Serial.print("pgm_read_byte( data ) = ");
	Serial.print( pgm_read_byte( data ) );
	Serial.println();
	
	Serial.print("pgm_read_byte( data + a ) = { ");
	for( int a=0; a < data_len ; a++ )
	{
		Serial.print( pgm_read_byte( data + a ) );
		Serial.print(", ");
	}
	Serial.println(" }");
		
	
	Serial.println("---"); // --- display data[ a ]
	
	Serial.print("data[ 0 ] = ");
	Serial.print( data[ 0 ] );
	Serial.println();
	
	Serial.print("data[ a ] = { ");
	for( int a=0; a < data_len ; a ++ )
	{
		Serial.print( data[ a ] );
		Serial.print(", ");
	}
	Serial.println(" }");
	
	
	Serial.println("---"); // --- display table[ 0x4000 +a ]
	
	Serial.print("data[ 0x4000 ] = ");
	Serial.print( data[ 0x4000 ] );
	Serial.println();
		
	Serial.print("data[ 0x4000 + a ] = { ");
	for( int a=0; a < data_len ; a ++ )
	{
		Serial.print( data[ 0x4000 + a ] );
		Serial.print(", ");
	}
	Serial.println(" }");
			
			
	Serial.println("---"); // --- display pgm_data[ a ]
	
	Serial.print("pgm_data[ 0 ] = ");
	Serial.print( pgm_data[ 0 ] );
	Serial.println();
	
	Serial.print("pgm_data[ a ] = { ");
	for( int a=0; a < data_len ; a ++ )
	{
		Serial.print( pgm_data[ a ] );
		Serial.print(", ");
	}
	Serial.println(" }");
	
		
	Serial.flush();
}

void loop()
{
}
````
